### ISSUE TITLE
FUNSD predictor

### ISSUE THREAD

**Issue by yudezhi123456** [NormXU/Layout2Graph#11]
I trained FUNSD-entity to get the best model, verified eval, and obtained entity-labeling consistent with your paper. However, the results of entity-Linking for 'Pair_F1_MACRO' and 'Pair_F1_MICRO' both reach 1.0. I wonder why? That's the metric, right? Below are screenshots from my yaml file, along with the experimental results.

![image](https://github.com/NormXU/Layout2Graph/assets/58250566/0792800a-3195-4f09-b756-590647d64445)

![image](https://github.com/NormXU/Layout2Graph/assets/58250566/c6c8c64b-3e4e-4c0a-9922-f7dd084c73c8)

![image](https://github.com/NormXU/Layout2Graph/assets/58250566/cc4aac39-3bb8-43c9-9e2a-71fe1ae313be)

![image](https://github.com/NormXU/Layout2Graph/assets/58250566/0978d3bb-5e59-4dad-8ab5-7541f583ec60)

**Comment by NormXU** [comment]
@yudezhi123456 
Yes, it is a kind of metric to reflect the performance regarding whether the model can predict the connections between cells correctly. 1.0 indicates that all cells have been predicted correctly.

**Comment by yudezhi123456** [comment]
> @yudezhi123456 Yes, it is a kind of metric to reflect the performance regarding whether the model can predict the connections between cells correctly. 1.0 indicates that all cells have been predicted correctly.

Thank you very much for your answer, There's a sentence In your paper "In the entity task,our model shown in Tab.2, achieved a F1 score of 0.80575 for entity-labeling and 0.77031 for entity-linking, using 32.98 million parameters."
Then I would like to know whether the F1 score of entity-labeling is expressed by two indicators: Node_F1_MACRO and Node_F1_MICRO, and which indicator is used to express the result of entity-linking, just like the value of 0.77031 mentioned in the paper.

**Comment by NormXU** [comment]
@yudezhi123456 
we report `Node_F1_MACRO` as the primary metric for entity labeling. Regarding entity linking, we use `Pair_F1_MACRO ` as the  metric in our paper. 


You can find the implementation details here:

evaluating entity-linking (Node_F1):

https://github.com/NormXU/Layout2Graph/blob/60586aa3956ab1598ff34de97cd3f68e0dc46adc/metrics/graph_layout_metrics.py#L52-L65

evaluating entity-labeling(Pair_F1):
https://github.com/NormXU/Layout2Graph/blob/60586aa3956ab1598ff34de97cd3f68e0dc46adc/metrics/graph_layout_metrics.py#L75-L87

**Comment by yudezhi123456** [comment]
> @yudezhi123456 we report `Node_F1_MACRO` as the primary metric for entity labeling. Regarding entity linking, we use `Pair_F1_MACRO ` as the metric in our paper.
> 
> You can find the implementation details here:
> 
> evaluating entity-linking (Node_F1):
> 
> https://github.com/NormXU/Layout2Graph/blob/60586aa3956ab1598ff34de97cd3f68e0dc46adc/metrics/graph_layout_metrics.py#L52-L65
> 
> evaluating entity-labeling(Pair_F1):
> 
> https://github.com/NormXU/Layout2Graph/blob/60586aa3956ab1598ff34de97cd3f68e0dc46adc/metrics/graph_layout_metrics.py#L75-L87

Thank you very much for your answer. I can understand what you mean and find the corresponding primary metric.
But after thinking about it all night, I still don't know which index entity-linking=0.77031 in the paper is obtained from. As you said, By "provide 'Pair_F1_MACRO' as the reported metric in our paper", but after I trained the FUNSD dataset and evaluated the test set, Pair_F1_MACRO=1.0, It is not consistent with 0.77031 in your paper. I can't find 0.77031 on Pair_F1_MACRO. Is it because I'm testing on a different dataset than yours? I'm testing testing-data from FUNSD, which is also generated by preprocess_data.py.
I am not sure whether I have made it clear. If you have time, I hope you can answer my confusion. Thank you very much.

**Comment by NormXU** [comment]
@weishu27 Could you please check out it and answer this questions?

**Comment by NormXU** [comment]
@yudezhi123456 It's quite weird that you can achieved a `Pair_F1_MACRO=1.0`. Could you please double-check that there is no data leakage from the testing dataset into the training dataset

Moreover, I notice from your log:
![Image_20231014110003](https://github.com/NormXU/Layout2Graph/assets/33339685/57b8d84a-31ab-4020-ac78-ddd40cd014c2)
It shows that your dataset only contains edges with label=0. There could be a potential bug with your data processing, as there should be edges with both label=0 and label=1. Label 0  represents an invalid connection, while label 1 denotes a valid connection.

**Comment by yudezhi123456** [comment]
> @yudezhi123456 It's quite weird that you can achieved a `Pair_F1_MACRO=1.0`. Could you please double-check that there is no data leakage from the testing dataset into the training dataset
> 
> Moreover, I notice from your log: ![Image_20231014110003](https://user-images.githubusercontent.com/33339685/275121037-57b8d84a-31ab-4020-ac78-ddd40cd014c2.png) It shows that your dataset only contains edges with label=0. There could be a potential bug with your data processing, as there should be edges with both label=0 and label=1. Label 0 represents an invalid connection, while label 1 denotes a valid connection.

I redownloaded the FUNSD dataset and examined the data and the test_convert_Funsd2Graph function, but still couldn't find the problem. The training data set consists of 149 samples and the test data set consists of 50 samples. Below is the structure of my dataset.

![image](https://github.com/NormXU/Layout2Graph/assets/58250566/0c0a1c76-4346-4bdf-aa1c-4217fdeee7ee)

Because when training the “word” model and verifying it, there are no such problems. Consistent with the paper, while training “entity”, the above problems occurred.
The following image shows the result of the “word” model validation.

![image](https://github.com/NormXU/Layout2Graph/assets/58250566/eff3f28f-0577-4820-a23d-50e4e18514c7)

![image](https://github.com/NormXU/Layout2Graph/assets/58250566/84537472-1f60-42f9-803f-12f485ffc544)

![image](https://github.com/NormXU/Layout2Graph/assets/58250566/4a0710c2-34ca-48ca-b2d1-86fd20429183)

I didn't make any unnecessary changes to your code, just the yaml file configuration.

The problem may not be in the data handling, it may be in the validation code itself. When I was training on the newly downloaded FUNSD, the following PairF1MACRO1.00000 appeared when I saved the model in the first round.

![image](https://github.com/NormXU/Layout2Graph/assets/58250566/bcf029f0-97ad-4ba8-934d-6c8b3a8dc3a5)


I am very confused, thank you for your patient answer.

**Comment by NormXU** [comment]
@yudezhi123456 Alright, I think I find out what causes this issue.

We provide two kinds of dataset and corresponding CollectFn as shown below:

<img src="https://github.com/NormXU/Layout2Graph/assets/33339685/f6042443-52d7-425a-b4a0-6a9d17077494" width="250">

The `GraphCollateFn` is for the node classification task. Consider a scenario where we are presented with a collection of OCR boxes. Each of these boxes is treated as an individual node within a graph. Our objective is to leverage the network to establish connections between certain nodes while disconnecting others. To achieve this, we initially implement a fully connected linkage strategy, but this approach proves overly dense for the network to learn from. As a remedy, we apply a sparser linkage approach based on geometric data and prior knowledge. The overlap between the fully connected and sparse linkages is then set as positive edges, while the rest edges within the sparse graph are set as negative edges.
Check the codes below:
https://github.com/NormXU/Layout2Graph/blob/ca627bfd3c1e3f1cfd0d5039fb031e396c01a5c5/networks/graph_net/graph_layout_net.py#L186-L195

However, if relational linking information is already available, as is the case with the FUNSD_entity_graph dataset,  we can directly use it. The dataset and collectFn in config file should be set to  `GraphLayoutEntityDataset `and `GraphEntityCollateF `respectively. Otherwise, all nodes will be set to negative edges based on our codes, which finally leads to the PairF1MACRO=1.0

In a word:
- For entity-node:
```yaml
datasets:
  train:
    dataset:
      type: GraphLayoutDataset
    collate_fn:
      type: GraphCollateFn

```
- For entity-linking
```yaml
datasets:
  train:
    dataset:
      type: GraphLayoutEntityDataset
    collate_fn:
      type: GraphEntityCollateFn
```
Sorry for the confusion. I will add the explanations in our documentation.

**Comment by yudezhi123456** [comment]
> @yudezhi123456 Alright, I think I find out what causes this issue.
> 
> We provide two kinds of dataset and corresponding CollectFn as shown below:
> 
> <img alt="" width="250" src="https://user-images.githubusercontent.com/33339685/275169555-f6042443-52d7-425a-b4a0-6a9d17077494.png">
> The `GraphCollateFn` is for the node classification task. Consider a scenario where we are presented with a collection of OCR boxes. Each of these boxes is treated as an individual node within a graph. Our objective is to leverage the network to establish connections between certain nodes while disconnecting others. To achieve this, we initially implement a fully connected linkage strategy, but this approach proves overly dense for the network to learn from. As a remedy, we apply a sparser linkage approach based on geometric data and prior knowledge. The overlap between the fully connected and sparse linkages is then set as positive edges, while the rest edges within the sparse graph are set as negative edges. Check the codes below:
> 
> https://github.com/NormXU/Layout2Graph/blob/ca627bfd3c1e3f1cfd0d5039fb031e396c01a5c5/networks/graph_net/graph_layout_net.py#L186-L195
> 
> However, if relational linking information is already available, as is the case with the FUNSD_entity_graph dataset, we can directly use it. The dataset and collectFn in config file should be set to `GraphLayoutEntityDataset `and `GraphEntityCollateF `respectively. Otherwise, all nodes will be set to negative edges based on our codes, which finally leads to the PairF1MACRO=1.0
> 
> In a word:
> 
> * For entity-node:
> 
> ```yaml
> datasets:
>   train:
>     dataset:
>       type: GraphLayoutDataset
>     collate_fn:
>       type: GraphCollateFn
> ```
> 
> * For entity-linking
> 
> ```yaml
> datasets:
>   train:
>     dataset:
>       type: GraphLayoutEntityDataset
>     collate_fn:
>       type: GraphEntityCollateFn
> ```
> 
> Sorry for the confusion. I will add the explanations in our documentation.

**Comment by yudezhi123456** [comment]
> @yudezhi123456 Alright, I think I find out what causes this issue.
> 
> We provide two kinds of dataset and corresponding CollectFn as shown below:
> 
> <img alt="" width="250" src="https://user-images.githubusercontent.com/33339685/275169555-f6042443-52d7-425a-b4a0-6a9d17077494.png">
> The `GraphCollateFn` is for the node classification task. Consider a scenario where we are presented with a collection of OCR boxes. Each of these boxes is treated as an individual node within a graph. Our objective is to leverage the network to establish connections between certain nodes while disconnecting others. To achieve this, we initially implement a fully connected linkage strategy, but this approach proves overly dense for the network to learn from. As a remedy, we apply a sparser linkage approach based on geometric data and prior knowledge. The overlap between the fully connected and sparse linkages is then set as positive edges, while the rest edges within the sparse graph are set as negative edges. Check the codes below:
> 
> https://github.com/NormXU/Layout2Graph/blob/ca627bfd3c1e3f1cfd0d5039fb031e396c01a5c5/networks/graph_net/graph_layout_net.py#L186-L195
> 
> However, if relational linking information is already available, as is the case with the FUNSD_entity_graph dataset, we can directly use it. The dataset and collectFn in config file should be set to `GraphLayoutEntityDataset `and `GraphEntityCollateF `respectively. Otherwise, all nodes will be set to negative edges based on our codes, which finally leads to the PairF1MACRO=1.0
> 
> In a word:
> 
> * For entity-node:
> 
> ```yaml
> datasets:
>   train:
>     dataset:
>       type: GraphLayoutDataset
>     collate_fn:
>       type: GraphCollateFn
> ```
> 
> * For entity-linking
> 
> ```yaml
> datasets:
>   train:
>     dataset:
>       type: GraphLayoutEntityDataset
>     collate_fn:
>       type: GraphEntityCollateFn
> ```
> 
> Sorry for the confusion. I will add the explanations in our documentation.

Thank you for your answer, which is of great help to the advancement of my research. I will continue to conduct research in this direction.

**Comment by CCball** [comment]
how to train on funsd dataset

**Comment by yudezhi123456** [comment]
> how to train on funsd dataset

According to the README given by the author, it is very clear that it can be trained step by step according to the requirements
